File: skip-iterator.js
// Time Complexity : O(N)
// Space Complexity : O(N)
// Did this code successfully run on Leetcode : Not a LeetCode question
// Any problem you faced while coding this : Just the seekCursor method was difficult to write. Everything else was fine.

class SkipIterator {
  constructor(list) {
    this.list = list;
    this.skipCount = new Map;
    this.cursor = 0;
  }

  hasNext() {
    return this.list[this.cursor] != null;
  }

  next() {
    let element = this.list[this.cursor];
    this.seekCursor();
    return element;
  }

  skip(num) {
    if (this.list[this.cursor] == num) this.seekCursor();
    else {
      if (this.skipCount.has(num)) {
        let newCount = this.skipCount.get(num) + 1;
        this.skipCount.set(num, newCount);
      } else {
        this.skipCount.set(num, 1);
      }
    }
  }

  seekCursor() {
    let element = this.list[++this.cursor];
    while (this.cursor < this.list.length && this.skipCount.has(element)) {
      let count = this.skipCount.get(element) - 1;
      if (count == 0) this.skipCount.delete(element);
      else this.skipCount.set(element, count);
      element = this.list[++this.cursor];
    }
  }
}

File: skip-iterator.cpp
// Time Complexity : O(mn)
// Space Complexity : O(n)
// Did this code successfully run on Leetcode : Yes

#include <iostream>
#include <unordered_map>
#include <vector>

class SkipIterator {
private:
    std::vector<int>::const_iterator it;
    std::unordered_map<int, int> count;
    int nextEl;

public:
    SkipIterator(std::vector<int>::const_iterator it) : it(it), nextEl(0) {
        advance();
    }

    bool hasNext() const {
        return nextEl != 0;
    }

    int next() {
        if (!hasNext())
            throw std::runtime_error("empty");
        int el = nextEl;
        advance();
        return el;
    }

    void skip(int num) {
        if (!hasNext())
            throw std::runtime_error("empty");
        if (nextEl == num) {
            advance();
        } else {
            count[num]++;
        }
    }

private:
    void advance() {
        nextEl = 0;
        while (nextEl == 0 && it != end) {
            int el = *it++;
            if (count[el] > 0) {
                count[el]--;
            } else {
                nextEl = el;
            }
        }
    }
};

int main() {
    std::vector<int> nums = {1, 2, 3};
    SkipIterator it(nums.begin());

    std::cout << it.hasNext() << std::endl;

    it.skip(2);
    it.skip(1);
    it.skip(3);

    std::cout << it.hasNext() << std::endl;

    return 0;
}
